### PR TITLE
made various corrections to the book

### PR DESCRIPTION
**Pull request by BartMassey**
Things still to do: 

* Add links or citations to all the technology mentioned in the book.
* Link section references to their destination.

### CODE REVIEW

File: doc/snek.adoc [hunk 1]
```diff
@@ -81,43 +81,47 @@ compiler running on a separate desktop or laptop system.
...
+   energy system that automatically tracked the sun. Mark is
+   now a mechanical engineering student at Washington
+   University in St Louis.
+
+The hardware was just what we wanted, and a few students
+used skills learned in the program later on. However, the
+software was not aimed at young students just starting to
+write code. Instead of throwing out our existing systems and
+starting over, I wondered if we couldn't keep using the same
+(hand-made) hardware and but improve the programming
```
Comment (keith-packard) [inline doc/snek.adoc]: and but -> but

File: doc/snek.adoc [hunk 2]
```diff
@@ -292,31 +298,33 @@ hello, world
...
+doesn't matter, but each line has to use the same
+indentation. When you're done with the list of statements,
+finish by going back to the old indentation level.
+
+While entering a list of statements, the command
 processor will prompt with `+` instead of `>` to let you know that
 it's still waiting for more input before it does anything. It's the
 “line with the old indentation level” that displays the second
-`+`. Hitting enter on that line ends the list of statements for the
+`>`. Hitting enter on that line ends the list of statements for the
```
Comment (keith-packard) [inline doc/snek.adoc]: Hrm. this is unclear (clearly). What I'm trying to explain is why you have to hit return twice after entering a compound statement in the repl:

> if true:
+  print('yup')
+ 
yup
>

File: doc/snek.adoc [hunk 3]
```diff
@@ -520,15 +534,16 @@ within the Tuple is followed by a comma:
...
-unchanging, which means Dictionaries can only be indexed by immutable
-values. Lists and Dictionaries are the only mutable data structures in
-Snek, so there are lots of options for Dictionary indices.(((Dictionary)))
+Dictionaries are the fanciest data structure in Snek. Like
+Lists and Tuples, Dictionaries hold multiple values. Unlike
+Lists and Tuples, Dictionaries are not indexed by
+numbers. Instead, Dictionaries are indexed by another Snek
+value. The only requirement is that the index value be immutable,
+so that it can never change. Lists and Dictionaries are the only mutable data
+structures in Snek, so there are lots of options for
```
Comment (keith-packard) [inline doc/snek.adoc]: 'so there are lots of options' doesn't seem related to 'the only mutable data structures in Snek'. Oh, I think what I was trying to say is that you can use anything other than a list or dictionary as a dictionary index. This is really unclear here.

File: doc/snek.adoc [hunk 4]
```diff
@@ -1061,14 +1098,17 @@ _value_ ``/`` _value_::
 The Divide operator performs division on numbers.(((/)))
 
 _value_ ``//`` _value_::
-The Div operator performs division on integer values, producing an
-integer result.(((//)))
+The Div operator performs "integer division" on numbers, producing a
+result that is the nearest integer less than or equal to the
+quotient.(((//)))
```
Comment (keith-packard) [inline doc/snek.adoc]: Should we add an explanation in terms of floor? a // b == floor (a/b)

File: doc/snek.adoc [hunk 5]
```diff
@@ -1105,10 +1145,10 @@ performs negation on numbers.(((–)))
 
 ``+`` _value_::
 When used as a unary prefix operator, the Unary Plus operator
-does nothing at all.(((+)))
+does nothing at all to a number.(((+)))
```
Comment (keith-packard) [inline doc/snek.adoc]: actually, unary + can be applied to any value as it's simply ignored in the grammar. I guess that could be considered a bug somehow?

File: doc/snek.adoc [hunk 6]
```diff
@@ -1185,7 +1226,7 @@ value for _bound_ is ``–1``. Here are some examples:
 
 String interpolation in Snek can be confused with formatted printing
 in other languages. In Snek, the ``print`` function takes a single
```
Comment (keith-packard) [inline doc/snek.adoc]: Hrm. the Snek print function takes as many values as you like. And fails to separate them with spaces. Ok, fixed that bug. Now I've got this for text:

    Print writes all of its positional parameters to the console separated
    by spaces (`' '`) followed by the `end` named parameter (default:
    `'\n'`).(((print)))

Comment (keith-packard) [reply doc/snek.adoc]: Hrm. 'to the console' is kinda buried in there. Where should it go?

File: doc/snek.adoc [hunk 7]
```diff
@@ -2081,14 +2125,14 @@ return -1 if no characters are pending.(((stdscr.getch)))
...
+The Snek Development Environment, Snekde, is a Python program which runs
 on Linux, Mac OS X and Windows to work with small devices running
 Snek, such as the Duemilanove and Metro M0 Express boards.(((snekde)))
 
 === Starting Snekde
 
-On Windows and Linux, launch snekde from your application menu. On Mac OS
-X, snekde is installed along with the other snek files in the Snek
+On Windows and Linux, launch `snekde` from your application menu. On Mac OS
+X, Snekde is installed along with the other Snek files in the Snek
```
Comment (keith-packard) [inline doc/snek.adoc]: `snekde` and Snekde in the same paragraph? Should those be the same?

File: doc/snek.adoc [hunk 8]
```diff
@@ -2415,8 +2459,8 @@ get the board into this mode by connecting the board to your computer
 over USB and then pressing the reset button twice in quick
 succession.
 
-Then, find the snek-feather-{version}.uf2 file included in the snek
-package for your machine and copy it to the CURRENT.UF2 file on the
+Then, find the `snek-feather-`{version}`.uf2` file included in the Snek
```
Comment (keith-packard) [inline doc/snek.adoc]: Hrm. {version} should be set in the same style as snek-feather here.

File: doc/snek.adoc [hunk 9]
```diff
@@ -2378,8 +2422,8 @@ get the board into this mode by connecting the board to your computer
 over USB and then pressing the reset button twice in quick
 succession.
 
-Then, find the snek-metrom0-{version}.uf2 file included in the snek
-package for your machine and copy it to the CURRENT.UF2 file on the
+Then, find the `snek-metrom0-`{version}`.uf2` file included in the Snek
+package for your machine and copy it to the `CURRENT.UF2` file on the
```
Comment (keith-packard) [inline doc/snek.adoc]: Hrm. {version} should be set in the same style as snek-metrom0 here.

File: doc/snek.adoc [hunk 10]
```diff
@@ -2459,8 +2503,8 @@ get the board into this mode by connecting the board to your computer
 over USB and then pressing the reset button twice in quick
 succession.
 
-Then, find the snek-crickit-{version}.uf2 file included in the snek
-package for your machine and copy it to the CURRENT.UF2 file on the
+Then, find the `snek-crickit-`{version}`.uf2` file included in the Snek
```
Comment (keith-packard) [inline doc/snek.adoc]: {version} should be set in the same style as snek-crickit

File: doc/snek.adoc [hunk 11]
```diff
@@ -2508,8 +2552,8 @@ to your computer over USB, sliding the switch to the right (towards
 the microphone) and then pressing the reset button twice in quick
 succession.(((Playground)))
 
-Then, find the snek-playground-{version}.uf2 file included in the snek
-package for your machine and copy it to the CURRENT.UF2 file on the
+Then, find the `snek-playground-`{version}`.uf2` file included in the Snek
```
Comment (keith-packard) [inline doc/snek.adoc]: {version} should be set in the same style as snek-playground

File: doc/snek.adoc (hunk 5 above)
Comment (BartMassey) [reply doc/snek.adoc]: Yeah, that's a bug. I want a syntax error when I say `+{1:2}` for sure.

File: doc/snek.adoc [hunk 12]
```diff
@@ -972,12 +1009,12 @@ you can store UTF-8 values in them comfortably. Just don't expect
...
-list, add or remove elements. In other languages, these are often
+Lists are ordered collections of values. You can change the contents of a
+list by adding or removing elements. In other languages, these are often
 called arrays or vectors. Lists are “mutable” values.(((List)))
 
 Tuples::
-Tuples are immutable lists of values. That is, you can't change the
+Tuples are immutable lists of values. You can't change the
 contents of a list once created, although if one of the elements of
 the list *is* mutable, you can modify that and see the changed results
```
[the remark this answers is not in the document]
Comment (BartMassey) [reply doc/snek.adoc]: Fixed.

File: doc/snek.adoc [hunk 13]
```diff
@@ -972,12 +1009,12 @@ you can store UTF-8 values in them comfortably. Just don't expect
...
-Lists are an ordered set of values. You can change the contents of a
-list, add or remove elements. In other languages, these are often
+Lists are ordered collections of values. You can change the contents of a
+list by adding or removing elements. In other languages, these are often
 called arrays or vectors. Lists are “mutable” values.(((List)))
 
 Tuples::
-Tuples are immutable lists of values. That is, you can't change the
+Tuples are immutable lists of values. You can't change the
 contents of a list once created, although if one of the elements of
```
[the remark this answers is not in the document]
Comment (BartMassey) [reply doc/snek.adoc]: Fixed.

File: doc/snek.adoc [hunk 14]
```diff
@@ -937,9 +974,9 @@ Here are some valid Dictionaries:
 { 1: 'one' }
 ----
 
-You can include entries with duplicate keys, the resulting Dictionary
+You can include entries with duplicate keys: the resulting Dictionary
```
[the remark this answers is not in the document]
Comment (BartMassey) [reply doc/snek.adoc]: Fixed earlier thing.

File: doc/snek.adoc [hunk 15]
```diff
@@ -893,13 +923,20 @@ Hex value. Use two hex digits to represent any character.
...
+Literal single-quote. Useful inside single-quoted strings.
 
 === List and Tuple Constants
 
-List and Tuple constants in Snek are values separated by commas
-enclosed in either square brackets (for Lists) or parentheses (for
-Tuples).
+List and Tuple constants in Snek are values separated by
+commas and enclosed in either square brackets (for Lists) or
+parentheses (for Tuples).
```
[the remark this answers is not in the document]
Comment (BartMassey) [reply doc/snek.adoc]: Fixed.

File: doc/snek.adoc [hunk 16]
```diff
@@ -850,13 +879,14 @@ separating out names and keywords from operators and other syntax.
...
 amount; statements in deeper blocks are indented more, statements in
-external blocks less.
+shallower blocks less.
 
-When typing Snek directly at the Snek prompt, blank lines become
-significant as Snek cannot know what you will type next. You can see
-this in the Tutorial where Snek finishes an indented block at the
+When typing Snek directly at the Snek prompt blank lines become
+significant, as Snek cannot know what you will type next. You can see
+this in the Tutorial, where Snek finishes an indented block at the
```
[the remark this answers is not in the document]
Comment (BartMassey) [reply doc/snek.adoc]: Yep. I think it's fine to mention it again here.

File: doc/snek.adoc [hunk 17]
```diff
@@ -666,15 +684,19 @@ upper 3
 other 5
 ----
 
+The `elif` statements try other alternatives if previous
+`if` tests have not worked. The `else` statement is tried if
```
[the remark this answers is not in the document]
Comment (BartMassey) [reply doc/snek.adoc]: Fixed. Went with "executed"

File: doc/snek.adoc (hunk 3 above)
Comment (BartMassey) [reply doc/snek.adoc]: Fixed.

File: doc/snek.adoc (hunk 6 above)
Comment (BartMassey) [reply doc/snek.adoc]: I think it's fine as you wrote it.

File: doc/snek.adoc (hunk 7 above)
Comment (BartMassey) [reply doc/snek.adoc]: Snekde is a component of Snek. I think it's fine.

File: doc/snek.adoc (hunk 8 above)
Comment (BartMassey) [reply doc/snek.adoc]: For all of these I think it's fine. I didn't want to work out how to fix the formatting, and it doesn't really matter.

File: doc/snek.adoc (hunk 7 above)
Comment (keith-packard) [reply doc/snek.adoc]: no, using `snekde` one place and Snekde in the other? Is that good?

File: doc/snek.adoc (hunk 16 above)
[the remark this answers is not in the document]
Comment (keith-packard) [reply doc/snek.adoc]: well, the trouble is that it's really unclear above, having confused you into thinking it should have said '> ' instead of '+ '. I think this text is fine, but the earlier text needs works.

File: doc/snek.adoc (hunk 5 above)
Comment (keith-packard) [reply doc/snek.adoc]: This would have to be checked at runtime (as Python3 does) to check the representation, not just catching a few cases in the parser where you use a constant. And that would require another op in the virtual machine, all to catch
errors in usage of an operator which otherwise doesn't do anything. Doesn't seem worth the space to me.

File: doc/snek.adoc [hunk 18]
```diff
@@ -315,9 +315,8 @@ finish by going back to the old indentation level.
 
 While entering a list of statements, the command
 processor will prompt with `+` instead of `>` to let you know that
-it's still waiting for more input before it does anything. It's the
-“line with the old indentation level” that displays the second
-`>`. Hitting enter on that line ends the list of statements for the
+it's still waiting for more input before it does anything. A
+blank line ends the list of statements for the
```
Comment (keith-packard) [inline doc/snek.adoc]: yeah, that's way better.

File: doc/snek.adoc [hunk 19]
```diff
@@ -540,8 +539,8 @@ Lists and Tuples, Dictionaries are not indexed by
 numbers. Instead, Dictionaries are indexed by another Snek
 value. The only requirement is that the index value be immutable,
 so that it can never change. Lists and Dictionaries are the only mutable data
-structures in Snek, so there are lots of options for
-Dictionary indices.(((Dictionary)))
+structures in Snek: anything else can be used as a
+Dictionary index.(((Dictionary)))
```
Comment (keith-packard) [inline doc/snek.adoc]: it's a bit more subtle than this -- you can't use a tuple containing a dict or list (the check is recursive), but I think this is probably fine

File: doc/snek.adoc [hunk 20]
```diff
@@ -935,8 +934,8 @@ Literal single-quote. Useful inside single-quoted strings.
 === List and Tuple Constants
 
 List and Tuple constants in Snek are values separated by
-commas and enclosed in either square brackets (for Lists) or
-parentheses (for Tuples).
+commas and enclosed in brackets: square brackets for Lists,
+parentheses for Tuples.
```
Comment (keith-packard) [inline doc/snek.adoc]: yeah. nice.

File: doc/snek.adoc [hunk 21]
```diff
@@ -1225,9 +1224,13 @@ as indicating _base_ and _bound_. Here are some examples:
 === String Interpolation
 
 String interpolation in Snek can be confused with formatted printing
-in other languages. In Snek, the ``print`` function takes a single
-String. String interpolation is how this String is generated from a format
-specification String and a List or Tuple of parameters.(((%)))(((string interpolation)))
+in other languages. In Snek, the ``print`` function takes String
```
Comment (keith-packard) [inline doc/snek.adoc]: actually, like python3, print takes *any* values. I fixed the 'print' reference, but missed this one.

**Comment by BartMassey** [comment]
We really can't be editing this at the same time. :-)

File: doc/snek.adoc (hunk 21 above)
Comment (BartMassey) [reply doc/snek.adoc]: Fixed.

**Comment by keith-packard** [comment]
I've rebased and merged this PR.
